src/network/network_crypto.cpp

```diff
@@ -0,0 +1,110 @@
+
+/**
+ * Call \c crypto_wipe for all the data in the given span.
+ * @param span The span to cryptographically wipe.
+ */
+static void crypto_wipe(std::span<uint8_t> span)
+{
+	crypto_wipe(span.data(), span.size());
+}
+
+template <size_t size>
+class SharedKeys {
+	std::array<uint8_t, 2 * size> keys;
+public:
+	~SharedKeys() { crypto_wipe(keys); }
+	constexpr std::span<const uint8_t> ClientToServer() const { return std::span(this->keys.data(), size); }
+	constexpr std::span<const uint8_t> ServerToClient() const { return std::span(this->keys.data() + size, size); }
+	constexpr std::span<uint8_t> Both() { return this->keys; }
+};
+
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+static bool X25519KeyExchange(SharedKeys<32> &shared_keys, const X25519Key &their_public_key,
+		const X25519Key &my_private_key, const X25519Key &my_public_key, std::string_view extra_payload)
+{
+	X25519Key shared_secret;
+	crypto_x25519(shared_secret.data(), my_private_key.data(), their_public_key.data());
+	if (std::all_of(shared_secret.begin(), shared_secret.end(), [](auto v) { return v == 0; })) {
+		/* Secret is all 0 because public key is all 0, just reject it */
+		return false;
+	}
+
+	crypto_blake2b_ctx ctx;
+	crypto_blake2b_init(&ctx, shared_keys.Both().size());
+	crypto_blake2b_update(&ctx, shared_secret.data(), shared_secret.size());
+	crypto_blake2b_update(&ctx, my_public_key.data(), my_public_key.size());
+	crypto_blake2b_update(&ctx, their_public_key.data(), their_public_key.size());
+	crypto_blake2b_update(&ctx, reinterpret_cast<const uint8_t *>(extra_payload.data()), extra_payload.size());
+	crypto_blake2b_final(&ctx, shared_keys.Both().data());
+	return true;
+}
+
+
+X25519AuthenticationHandler::X25519AuthenticationHandler()
+{
+	RandomBytesWithFallback(this->my_private_key);
+	crypto_x25519_public_key(this->my_public_key.data(), this->my_private_key.data());
+
+	RandomBytesWithFallback(this->nonce);
+}
+
+X25519AuthenticationHandler::~X25519AuthenticationHandler()
+{
+	crypto_wipe(this->nonce);
+}
+
+/* virtual */ void X25519AuthenticationHandler::SendRequest(Packet &p)
+{
+	p.Send_bytes(this->my_public_key);
+	p.Send_bytes(this->nonce);
+}
+
+std::optional<std::string> X25519AuthenticationHandler::DoReceiveResponse(Packet &p, std::string_view shared_key_extra_payload)
+{
+	X25519Key their_public_key{};
+	X25519Key message{};
+	std::array<uint8_t, 16> mac; // message authentication code
+
+	p.Recv_bytes(their_public_key);
+	p.Recv_bytes(message);
+	p.Recv_bytes(mac);
+
+	SharedKeys<32> shared_keys;
+	if (!X25519KeyExchange(shared_keys, their_public_key, this->my_private_key, this->my_public_key, shared_key_extra_payload)) return {};
+
+	if (crypto_aead_unlock(message.data(), mac.data(), shared_keys.ClientToServer().data(),
+			nonce.data(), their_public_key.data(), their_public_key.size(), message.data(), message.size()) != 0) return {};
+
+	if (crypto_verify32(message.data(), shared_keys.ServerToClient().data()) != 0) return {};
```

Comment:
Does this means the client is sending his shared key over the wire?

The idea of those shared keys is that they are never (ever) exchanged over the wire, encrypted or not. Similar as a private key. They need to remain local.
So even if we know this packet is just to have something in there, the public key you had before is better than this :)

Maybe for now just send a single byte, a `1` or something, to not give the impression it does anything else?

As I mentioned in another comment, I would suspect we use this as a start to encrypt the traffic (as we might as well, we are all the way there after all). So it might change later anyway.

Reply:
Yeah, I was already rethinking what to do instead. JGRPP has some incrementing counter, which could work although for now there would be very little to test content wise. Ah well, I'll keep thinking ahead for a while to find a nice solution.

Reply:
Don't overthink this. The payload is not relevant for what you try to achieve here :)

A counter only works for a connection that remains open, as how JGRPP uses it. So you can't try a company password twice with the same payload. But as this comes from a new connection, that doesn't really solve anything.

Reply:
It's definitely true that the payload is not relevant here.

However, it would be nice if the decrypt that we are intending to do at some later point, were to be able to use the exact same code in OpenTTD (and as such the libraries). And since it's not the game connection, but the admin connection, it's (much) harder to make changes at a later stage without breaking things.

Reply:
It is also why I would suggest we do this in 15.0, to give ourselves a bit of time figuring out what to do here exactly.

But yeah, creating something that survives the future is the tricky part. Maybe give a run out multiplayer authentication and encryption, and see how that fits in this work?

Reply:
I'm toying with the idea of a quite surgical approach to introducing encryption of the data. I'm not sure whether it's a good idea yet, but why not encrypt the payload of "all" the packets that we send over the line?
Or at least those packets when we have a shared secret. This would put the nonce and mac (signature) between the size of the packet and the type byte of the packet. That will increase each packet by 40 bytes though.

I was wondering whether sending the packet size unencrypted was useful, but I doubt it makes any sense not to send it unencrypted because the decrypt needs to know the length. And that we can derive from the packet size.

The major question is whether adding 40 bytes to the packet is a big problem. Also, do we want packet sequence numbers to prevent replay attacks? That would add another 4 or 8 bytes.
Although we could be sneaky, as there is no real reason to make a whole new random nonce each time. It is, as far as I can see, fine to just increment the value by one each iteration. The monocypher documentation even mentions that for shorter nonces random is not safe enough, and you should use a counter. So, in theory we need to agree on a nonce once (say 16 bytes), and just append a 8 byte counter to that, that is incremented with each packet. That would implicitly prevent replay attacks, as the nonce would be wrong. In other words, with just 16 bytes per packet, less than the size bump for going from IPv4 to IPv6, we could encrypt every packet after the key exchange.

The only question is then, at what point do we want to perform the jump to encryption.  For game I'd say after the NewGRF and client version have been checked, but before the name is sent. That way old clients attempting to join just get the "right" you've got the wrong version error, instead of some invalid/illegal packet message. Since we already know the version is "right", the client can also put their public key in the join packet. If the server then sends its public key and starter nonce back, both sides have enough information to encrypt the second packet that either client and server sends.

For admin we got the method handshake, so the client sends the methods it supports. The server chooses and and sends the nonce and public key, and the the client replies with its public key and everything can be encrypted from then on.

There seems to be a limit of [256 GiB](https://pycryptodome.readthedocs.io/en/latest/src/cipher/chacha20_poly1305.html) for the maximum amount of plaintext before a new key must be agreed upon. That is a heck of a lot, but with maximum packet size that's "only" 8 million packets (2**23). On the other hand... that's an insane amount, so just "oops"-killing the connection after 256 GiB of packets per socket might not even be the worst solution.

Reply:
40 bytes per packet sounds fine by me. About the same overhead as the IP header itself.

Otherwise, sounds like a plan to at least try out! We won't get this right first drafts, and there is nothing obvious weird in your story, so yeah.

About the nonce, replay attacks aren't actually possible. Monocypher cycles the key every packet send, and the nonce is random at start. So this is already sequenced. At least, as far as I can tell, there isn't a packet you can replay and would be accepted at any stage, assuming the server generates the initial nonce. That is the lovely thing about these methods, they are actually secure with little effort :)